### PR TITLE
Upgrade to `emulator 0.15.0`

### PR DESCRIPTION
**Pull request by stephenctw**
Tests and CI would require the latest test suite from [machine-emulator ](https://github.com/cartesi/machine-emulator)which is not yet officially released.
Implements #41, #45 and #48 .

### CODE REVIEW

File: contracts/AccessLogs.sol
Comment (GCdePaula) [inline contracts/AccessLogs.sol:49]: I believe this should be the index of the leaf, rather than the memory address. A leaf (and its index) represents a word; an address represents a byte.

I also think the type here could be `Memory.Stride`.

File: contracts/AccessLogs.sol
Comment (GCdePaula) [inline contracts/AccessLogs.sol:52]: I think we should always avoid `wrap` and `unwrap` outside the `Memory` library. I think the library should expose enough constructors and destructors that we don't ever need to use `wrap` and `unwrap`.

File: contracts/AccessLogs.sol
Comment (GCdePaula) [inline contracts/AccessLogs.sol:60]: I'm not sure this would have issues with the rest of the code, but I also think this should be `Memory.Address`.

**Review by GCdePaula — COMMENTED** [review]
Great! That's the direction I had in mind, looking good! I have a few comments.

I'm unsure of the amount of checking between `Stride`, `Address`, `AlignedSize`, and `Region` we should do.

For example, stride can never be larger than `2^61 - 1`, and `AlignedSize` can never be larger than `61` (log2 of the number of leafs). Furthermore, depending on the value of `AlignedSize`, the maximum value of `Stride` can be even smaller; for example, when aligned size is 61, stride has to be zero, and when it is `60` it can be either `0` or `1`. The only case stride can have the maximum value of `2^61 - 1` is when aligned size is `0`.

The question is if we should design our constructors in a way that it's impossible to build these values with wrong arguments. I believe we'll answer this question with time. After each iteration, we should ask "where can a user of these libraries pass wrong values in a way that would cause a bug undetected".

Additionally, feel free to change any naming, specially on the constructors/destructors.

File: contracts/AccessLogs.sol
Comment (stephenctw) [reply contracts/AccessLogs.sol:60]: How about I wrap the type `Memory.Address` in `UArchCompat` library?

File: contracts/Memory.sol
Comment (stephenctw) [reply contracts/Memory.sol:49]: How about `Memory.Position` or `Memory.StartPosition`? It's probably not ideal either but we used to interchangeably use `paddr` and `position` for the same variable before.

File: contracts/MetaStep.sol
Comment (stephenctw) [reply contracts/MetaStep.sol:36]: My bad, I assumed the meta-counter is the same as the micro-cycle value.

File: contracts/UArchState.sol
Comment (stephenctw) [reply contracts/UArchState.sol:19]: Yeah this is not a good coding practice. The reason behind this is to have some flexibility over the `UArchState` implementation. We use another implementation (`test/UArchStateAux`) to perform the binary test for the `step`. And having a single library with all the functions loses this flexibility, because importing a library is static. That's why we need an interface and its inheritance. Do you have better suggestion for this problem?

File: contracts/interfaces/IAccessLogs.sol
Comment (stephenctw) [reply contracts/interfaces/IAccessLogs.sol:17]: Do you suggest moving it to another interface or library? If we move it to `IMetaStep` or `IUArchStep`, it may not make sense to import the interface in the `AccessLogs` library. If we move it to the `AccessLogs` library itself, any contract that want to use the structure will spend more deployment cost because all library functions are internal. What do you think?

File: contracts/MetaStep.sol
Comment (stephenctw) [reply contracts/MetaStep.sol:24]: Yeah I can try.

File: contracts/Memory.sol
Comment (GCdePaula) [reply contracts/Memory.sol]: I like them both!

File: contracts/UArchState.sol
Comment (GCdePaula) [reply contracts/UArchState.sol:20]: Ah, I see it now.

Hmm, maybe it's a silly idea. But can't we have multiple "profiles" in the `foundry.toml` configuration file?

I'm thinking that we have different remappings there; one for testing stuff the way you described, where the import resolves to the test library, and another for the real thing, where the import resolves normally. Not sure it works.

I don't know, maybe it's too crazy. Feels a bit hacky haha. But I feel we wanted to write things one specific way, but Solidity gets in the way. Thus we sacrificed functionality a bit so that we can test things properly. Maybe this profile/remappings idea (if it works) allows us to code closer to what we wanted.

File: contracts/interfaces/IAccessLogs.sol
Comment (GCdePaula) [reply contracts/interfaces/IAccessLogs.sol:17]: I think when all methods are internal, the compiler will strip all unused/unreachable/dead code. But I'm not sure! Do you know?

File: contracts/Memory.sol
Comment (stephenctw) [reply contracts/Memory.sol]: I'm going for `PhysicalAddress`

File: contracts/interfaces/IAccessLogs.sol
Comment (stephenctw) [reply contracts/interfaces/IAccessLogs.sol:17]: You are right! I verified the deployment cost didn't increase at all. So I'll move the struct back to `AccessLogs` library.

File: contracts/UArchState.sol
Comment (stephenctw) [reply contracts/UArchState.sol:20]: Indeed it's a good idea. I'll experiment with that. Ideally we'll create a `foundry_test.toml` and pass it to the foundry/forge command when we are running the tests/coverage.

Comment (stephenctw) [reply contracts/UArchState.sol:20]: I think it's not feasible after several attempts. The main problem is that even in the test we need both implementations. We need normal `AccessLogs` to run `Replay` tests and test the library itself; We need hacked `AccessLogs` to run `Binary` tests. When the profile is selected it applies to all contracts, but what we need here is more like a per file run-time flexibility allowing us to freely determine which implementation to use.

File: contracts/MetaStep.sol
Comment (stephenctw) [reply contracts/MetaStep.sol:24]: So I turned everything into libraries except `UArchState` and `MetaStep`. Mainly because we need the flexibility of using different `AccessLogs` for testing, and not exposing this issue to regular users as described in #43

File: contracts/UArchState.sol
Comment (GCdePaula) [reply contracts/UArchState.sol:20]: Ah, what a shame.

Even if we have multiple profiles? One for replay tests and another for binary tests?

I think we can configure different ignores for different profiles, so each profile would have its own remappings and would only run the few specific tests it needs for that specific remapping.

Maybe we'd need to split the test file into multiple, I don't know. Maybe not.

If you need both "libraries" at the same time on the same test case execution, I think there's a hack here. We could on one place import X and on another import Y. In the deployment profile, both X and Y would resolve to the same file. On the multiple test profiles, X and Y could resolve to different things.

This comment is a bit all over the place; I just wanted to share all these tricks and tools we have at our disposal.

I think it would be a nice thing to have everything being a library, but if it's not possible or if it's too hacky, then we keep things as they are.

Comment (stephenctw) [reply contracts/UArchState.sol:20]: I think it's not only about the tests execution. There's no problem to disable a certain tests with a (non-)matching pattern. But during the `forge build`, it already imported the library code and compiles all the contracts. In order to swap the library import seamlessly, the library names and function signatures have to be the same so we don't have to change any caller contract/library code. But again, it doesn't seem that `foundry` is supporting this conditional compilation as I seen in the [issue](https://github.com/foundry-rs/foundry/issues/4709).

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:43]: Is this function only used in tests? If so, should we make it available only for tests?

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:28]: I'm thinking, should we make the pair buffer and pointer their own type? Something like:

```
struct Buffer {
    bytes data;
    uint64 pointer;
}

struct Context {
    bytes32 currentRootHash;
    Buffer buffer;
}
```

I think it would make things easier. Could even be on its own library.

Also, any special reason to user a `uint128`?

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:52]: If we go for a buffer type, this function would receive a buffer. We could call it like this:

```
a.buffer.toBytes32()
```

Also, I think we can name this function better. What do you think of `readBytes32` and `peekBytes32`?

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:77]: If we create a type for `Buffer`, we could create a method like `consume32Bytes` (or a better name), and call it instead of this line. Could be a function that both returns a `bytes32` and advances the pointer, or a function that just advances the pointer (and we read like we're already reading).

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:80]: Likewise for this one: `consumeBytes32(uint)` (or a better name). Also, although I'm pretty sure Solidity compiler already does this, multiplying by `32` can be achieved by shifting `5`.

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:103]: Likewise for these two lines: they could be a single `consumeWord` method on `Buffer` type.

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:104]: Are you swapping twice?

Could you just swap once on the return statement? Then it looks symmetric: read swaps just before returning, and write swaps just as it's beginning.

Comment (GCdePaula) [reply src/AccessLogs.sol:104]: Furthermore, I think we should avoid typing integers of the "wrong" endianness as integers. They should be `bytes8`. I think the correct type of a word is `bytes8`.

I think the signature of `uint64SwapEndian` should be `(bytes8 num) -> uint64`. This way I think we'll never mix things up.

If we create the type `Buffer`, a function that reads a word should return `bytes8`. Right at the end of this `readWord` function, we call swap, and it will become a `uint64`.

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:167]: These two arguments are the new `Buffer` type. If we create it, do you think maybe this could be a method of the `Buffer` type? I don't know if it makes sense.

**Review by GCdePaula — COMMENTED** [review]
Looking really good! I have some comments. I haven't finished reading everything yet; I'll add more reviews later if I have more comments.

File: src/AccessLogs.sol
Comment (stephenctw) [reply src/AccessLogs.sol:28]: The pointer was type `uint128` in the first version you wrote. I didn't change it. Are you suggesting to use `uint64` or `uint256`?

File: src/AccessLogs.sol
Comment (stephenctw) [reply src/AccessLogs.sol:43]: Not sure, because all the tests are compiled as the source code in the same time.

File: src/AccessLogs.sol
Comment (stephenctw) [reply src/AccessLogs.sol:167]: I think we'll need two functions, one to be `getRootWithoutPointerAdvanced`, the other one to be `getRootWithPointerAdvanced`. Because in `writeRegion` method, we have to `getRoot` twice, the first one being validation, the second one to calculate new hash. And the pointer cannot be advanced in between the calls.

File: src/AccessLogs.sol
Comment (stephenctw) [reply src/AccessLogs.sol:104]: Maybe it's better to be `bytes8 -> bytes8`?

File: src/AccessLogs.sol
Comment (GCdePaula) [reply src/AccessLogs.sol:43]: Insana thinking more of putting it in a auxiliary test file than on the main library code.

File: src/AccessLogs.sol
Comment (GCdePaula) [reply src/AccessLogs.sol:28]: I think either one is fine. A `uint64` is enough, but I think it won't occupy less space than a `uint256` because of padding. And I think a `uint256` uses less gas because it's the size of a word in the EVM.

So I think we should use `uint256`, and if we add more stuff to the struct later that would make a `uint64` take less space, we should consider changing.

What do you think?

File: src/AccessLogs.sol
Comment (GCdePaula) [reply src/AccessLogs.sol:104]: Yeah, I think that's also ok! But I don't know if the step function needs read to return an integer. Like, if the C++ version does everything using integers (and so does our translated Solidity version), then I think read must always return an integer. If that's not the case, I think returning a `bytes8` is ok!

File: src/AccessLogs.sol
Comment (GCdePaula) [reply src/AccessLogs.sol:167]: Yeah, hadn't thought of that! I like it.

File: src/AccessLogs.sol
Comment (stephenctw) [reply src/AccessLogs.sol:43]: Oh that makes sense, I'll change it.

File: helper_scripts/generate_AccessLogs.sh
Comment (GCdePaula) [reply helper_scripts/generate_AccessLogs.sh:23]: Another interesting thing is that a Solidity file with no directives that goes through the preprocessor will output exactly the same file. We could always invoke the preprocessor for all Solidity files.

Additionally, I wouldn't say the access logs file is a "template" file. The file we have for step is a template. Like, while we are developing the step code, we won't really be editing the template file this much; we'll be editing the generation script and the original C++ code. On the other hand, I see the AccessLogs differently.

Joining these two ideas together, an interesting design is to have files like `AccessLogs.sol.template` be normal Solidity files, instead of having a special name. So there would be just one `AccessLogs.sol` file, which is valid Solidity and has preprocessor directives. This file we put on `src/`.

Then we can create another directory, called `ready_src/` or something. We configure forge to build using this new directory. Our `makefile` will run the preprocessor through all files in `src/`, and fill the `ready_src/` with the output. I feel this way it will be easier to develop since our source files are really Solidity files, and it will be clearer which files are editable and which are not.

Just an idea! Not sure it makes sense.

Just reiterating that there are still template files, like the step one. If we use this idea, we would the step Solidity file using the template, and put it on `src/`. Then we'd run the preprocessor on it, which would do nothing.

**Review by GCdePaula — COMMENTED** [review]
I took a careful read through the `AccessLogs` generation, and I left a few comments.

File: src/Buffer.sol
Comment (stephenctw) [reply src/Buffer.sol:61]: The `32` is added because we are accessing a `byte` array. And an array in solidity always starts with its length which is a 32 byte-long variable. I'll put this in document.

File: helper_scripts/generate_AccessLogs.sh
Comment (stephenctw) [reply helper_scripts/generate_AccessLogs.sh:23]: The script basically deletes one block (production or mock) using `sed` depending on which mode it's being executed with.

In the first place I thought it'd be nice to have one single mechanism (bash) to generate all files that need extra work, i.e. `Step, AccessLogs, Constants`.... If the user understand one script, he'll probably understand all scripts easily.  But indeed you are making a good point that these two things aren't much the same. (Step and AccessLogs) And it can increase complexity for maintenance and readability as the code base grows. I'll switch to the way you suggested. Now it makes more sense to me.

BTW, do you have problem understanding the other two scripts? I'm worried it may also be hard to read for a normal user.

Comment (GCdePaula) [reply helper_scripts/generate_AccessLogs.sh:21]: I think the biggest issue for me personally is that I'm not great with bash and its idiosyncrasies. Like, carefully reading the script and knowing what you wanted to do, I can kinda infer what's going on. But I could never catch a bug in it, or catch some hidden pitfall. Or even tell if it's going to work just by looking at it. Using a bash script just turning blocks on and off I think it's too much.

For scripts that do the C++ -> Solidity conversion, I think they're good. I think for the constants it would be possible to use preprocessor macros (we invoke `gpp` defining the constants with `-D`) but I don't think it's a good idea. I prefer what you're doing.

Reading the step generation script, it's clear to me that we have a Solidity template, a C source, and we do a series of substitutions using `sed` to get a Solidity source file. The hard part of understanding this entire pipeline is not on bash, and we can easily abstract this generation and treat the step as a black box.

So I don't think the other generation scripts are a problem! They're harder to understand, but it's because the whole pipeline is hard (C++ step code, written in a weird way to be translatable, then passed by a series of substitutions). The scripting part of the substitutions are not the hardest part.

**Review by GCdePaula — COMMENTED** [review]
Couple more comments!

File: src/AccessLogs.sol
Comment (GCdePaula) [inline src/AccessLogs.sol:14]: Do you think it's possible to put this line in a Solidity comment? I think the preprocessor will still see it and do substitutions, and the file will remain valid Solidity. What do you think?

File: Makefile
Comment (GCdePaula) [reply Makefile:44]: Maybe also run `forge clean`?

File: foundry.toml
Comment (GCdePaula) [inline foundry.toml:11]: Any reason for skipping `src/`? Like, I believe if we make all files valid Solidity, `fmt` should work.

File: ready_src/Buffer.sol
Comment (GCdePaula) [inline ready_src/Buffer.sol:31]: Do you think it makes sense to create a `peekBytes8`?

Reusing the `peekBytes32` and doing a typecast makes me want to double check haha. Creating a specific bytes8 peek functions allows you to still reuse the `peekBytes32`, but have a specific test for that which will remove all doubts.

Also, this code we've never done an integration test, right? Like, does the whole replay test and other similar tests use this buffer?

I'm thinking on how we'll build this buffer offchain. If we're not doing an integration test using the buffer, maybe there's also a way to create an integration test for it like the replay. Let's think about it; maybe you can teach me how we're generating the replay tests, and we can do a similar thing here (if we're not already doing it!).

File: .gitignore
Comment (GCdePaula) [inline .gitignore:10]: Should we also ignore the whole `ready_src/` dir?

File: foundry.toml
Comment (stephenctw) [reply foundry.toml:11]: I thought the function conflicts in `production` and `mock` can fail the `fmt` command. I was wrong, my bad. Will not sKip `src`

File: ready_src/Buffer.sol
Comment (stephenctw) [reply ready_src/Buffer.sol:31]: The `replay` tests use the buffer.

File: .gitignore
Comment (stephenctw) [reply .gitignore:10]: One thing nice to have the `ready_src` committed is that when someone wants to use this project as a submodule, it' ready right away. If we ignore the directory, he'll have to generate it manually. What do you think?

Comment (GCdePaula) [reply .gitignore:10]: Good point! On the one hand, it's generated stuff (just run `make ...`). It feels like good practice not adding it to git. On the other hand, it can be quite practical. I don't know haha, I'll let you decide :)

File: ready_src/Buffer.sol
Comment (GCdePaula) [reply ready_src/Buffer.sol:31]: If they use it that's perfect! I thought they could be using the mocked stuff, which didn't include the buffer. Ignore the second part of my comment haha.

File: .gitignore
Comment (stephenctw) [reply .gitignore:10]: I think I'll keep `ready_src/` checked-in in github. So when people are trying to using this library, they simply add it as a submodule, and don't have to worry about dependencies like `emulator, Make, Docker`...
Also, if the person accidentally modified the library source code, (either manually or with script), it may be hard to debug without git telling that the source tree has been changed.

**Comment by stephenctw** [comment]
Changed the PR title to "Upgrade to `emulator 0.15.0`". The `reset` feature takes longer than expected. This PR contains only the "framework" of `reset`, not the actual implementation. Many features and changes are related to `emulator 0.15.0`, so it makes sense to merge this PR so it and be released along with SDK and `emulator 0.15.0`.

File: README.md
Comment (stephenctw) [reply README.md:63]: I tried `cpp`, however it lacks some feature flexibility that `gpp` offers. This can be a future improvement but I am afraid we don't have time to change this for this release.

File: Makefile
Comment (edubart) [inline Makefile:25]: The CI needs a check to know whether a regeneration is needed, because the emulator sources may change and we will may forget to manually regenerate the sources in `ready_src`. But lets leave this as an enhancement for later.

File: Makefile
Comment (edubart) [inline Makefile:26]: I am confused which task will regenerate `ready_src`, I could only regenerate it with `make all`, while doing `make generate-all` after `make all` revert `ready_src` back.

And if I do not regenerated it I got the following test errors:

```
Running 7 tests for test/AccessLogs.t.sol:AccessLogsTest
[PASS] testEndianSwap() (gas: 1797)
[PASS] testReadWord() (gas: 169743)
[FAIL. Reason: Call did not revert as expected] testReadWordRoot() (gas: 172653)
[FAIL. Reason: Call did not revert as expected] testReadWordValue() (gas: 172515)
[PASS] testWriteWord() (gas: 169240)
[FAIL. Reason: Call did not revert as expected] testWriteWordRootPosition() (gas: 172278)
[FAIL. Reason: Call did not revert as expected] testWriteWordRootValue() (gas: 175815)
Test result: FAILED. 3 passed; 4 failed; 0 skipped; finished in 4.31ms

Running 1 test for test/Memory.t.sol:MemoryTest
[PASS] testStrideAlignment() (gas: 497636309)
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 62.26ms
Ran 3 test suites: 6 tests passed, 4 failed, 0 skipped (10 total tests)

Failing tests:
Encountered 4 failing tests in test/AccessLogs.t.sol:AccessLogsTest
[FAIL. Reason: Call did not revert as expected] testReadWordRoot() (gas: 172653)
[FAIL. Reason: Call did not revert as expected] testReadWordValue() (gas: 172515)
[FAIL. Reason: Call did not revert as expected] testWriteWordRootPosition() (gas: 172278)
[FAIL. Reason: Call did not revert as expected] testWriteWordRootValue() (gas: 175815)
```

**Review by edubart — CHANGES_REQUESTED** [review]

**Review by GCdePaula — COMMENTED** [review]
The Solidity code around `step` looks great to me! I'm having trouble running the entire thing.

**Review by edubart — CHANGES_REQUESTED** [review]

File: ready_src/UArchStep.sol
Comment (edubart) [reply ready_src/UArchStep.sol:1]: Ok, I noticed this was another issue I pointed in comment https://github.com/cartesi/machine-solidity-step/pull/46#discussion_r1294911895

File: src/UArchCompat.sol
Comment (edubart) [reply src/UArchCompat.sol:1]: Ok, I noticed it's because you are copying existing `src` files to `ready_src` plus generated files. Is it possible to reorganize to not have duplicated code? I don't mind duplicated code of template files, but looks like some files are being copied without any change, this is confusing, is everything in `src` a template now?

File: Makefile
Comment (edubart) [reply Makefile:45]: Very odd because `make build test-all` fails here, while `make build && make test-all` works here. Any idea?

File: Makefile
Comment (stephenctw) [reply Makefile:25]: Yeah @mpernambuco mentioned this too. Maybe use `shasum` to check?

File: src/UArchCompat.sol
Comment (stephenctw) [reply src/UArchCompat.sol:1]: The logic behind these three directories "templates", "src" and "ready_src" is like below
1. "template files" are processed by scripts, and can produce code in "src" and "test"
2. "src files" will then be processed by `gpp` to handle macros, and then output to "ready_src"

The duplication you're seeing is because some files don't have macros to be processed. Since `gpp` has no problem bypassing those files, we put the files in "src" even if it needs no "preprocessing".

File: Makefile
Comment (stephenctw) [reply Makefile:45]: It's that the `generate-prod` are called twice in `make all`, but second occurrence is ignored by `make`. I'll rearrange the targets to not have this situation.

**Comment by edubart** [comment]
I could run the tests locally and made a quick overview, I will approve when the CI pass (after emulator prerelease). For other reviewers please note that this time I did not do a thorough review of the Solidity code.

**Review by GCdePaula — DISMISSED** [review]
Managed to run everything, all tests working :)

Solidity code around step looks good to me. Let's way for @edubart's approval.

**Review by edubart — APPROVED** [review]

**Review by mpernambuco — APPROVED** [review]

**Pull request merged**
